Possible reference notebook:
- https://eugejoh.netlify.app/post/wm-breaks/
- https://github.com/eugejoh/TO_Watermain

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
break_data = pd.read_csv("./data/Jan21_Update_Water_Main_Breaks.csv")
break_data.sample(10)

X          Y  OBJECTID  WATBREAKINCIDENTID  \
2520 -80.528681  43.446317     22721              193557   
744  -80.460568  43.425233      8729                2339   
822  -80.516441  43.427150      8807                 604   
57   -80.479430  43.462667      8042                1847   
572  -80.454928  43.430736      8557                1958   
653  -80.495565  43.449513      8638                  30   
1864 -80.519117  43.429929      9849                2257   
1407 -80.453352  43.439260      9392                 400   
427  -80.501131  43.453445      8412                  11   
74   -80.513022  43.445058      8059                 339   

               INCIDENT_DATE BREAK_TYPE ROAD_CLOSED SIDEWALK_CLOSED  \
2520  2020/02/27 07:42:39+00       MAIN      Closed            Open   
744   2018/01/20 00:00:00+00       MAIN        Open            Open   
822   2006/03/23 00:00:00+00       MAIN        Open            Open   
57    2013/02/23 00:00:00+00       MAIN        Open            Open   
572   2014/01/26 00:00:00+00       MAIN        Open            Open   
653   2000/02/14 00:00:00+00       MAIN        Open            Open   
1864  2017/02/15 00:00:00+00       MAIN        Open            Open   
1407  2003/02/02 00:00:00+00       MAIN        Open            Open   
427   1999/07/07 00:00:00+00       MAIN        Open            Open   
74    2004/02/26 00:00:00+00       MAIN        Open            Open   

     HOUR_IMPACTED UNITS_IMPACTED  ...  CIVIC_NUMBER           STREET ASSETID  \
2520    8-12 hours           0-50  ...            30    WESTWOOD CRES   42890   
744     8-12 hours            NaN  ...           270      KIPLING AVE   39050   
822     8-12 hours            NaN  ...           160        CYPRUS DR    8780   
57      8-12 hours            NaN  ...           547  WELLINGTON ST N   41690   
572     8-12 hours            NaN  ...           156       FOURTH AVE   15200   
653     8-12 hours            NaN  ...            69       WATER ST S   76027   
1864    8-12 hours            NaN  ...          1430      QUEENS BLVD   31670   
1407    8-12 hours            NaN  ...            30      KENNETH AVE   21030   
427     8-12 hours            NaN  ...           575        KING ST W   76569   
74      8-12 hours            NaN  ...           415    BELMONT AVE W    2010   

      ASSET_DEPTH FROST_DEPTH ASSET_SIZE ASSET_YEAR_INSTALLED ASSET_MATERIAL  \
2520          NaN         NaN      150.0               1966.0             CI   
744           NaN         NaN      150.0               1959.0             CI   
822           NaN         NaN      150.0               1963.0             CI   
57            NaN         NaN      300.0               1929.0             CI   
572           NaN         NaN      150.0               1960.0             CI   
653           NaN         NaN      300.0               2007.0            PVC   
1864          NaN         NaN      200.0               1966.0             DI   
1407          NaN         NaN      150.0               1958.0             CI   
427           NaN         NaN      250.0               1889.0             CI   
74            NaN         NaN      300.0               1957.0             CI   

      ASSET_EXISTS                              GLOBALID  
2520             Y  f8306243-8bcf-47f0-b5a6-431031431e23  
744              N  ce7a335a-2633-47e5-bf75-2fa02e86359f  
822              Y  5f566e78-5e00-42d4-9018-54173c339ddf  
57               Y  8322c43b-63a9-4077-8f44-a83f7bb117a9  
572              Y  f42f577e-850b-406f-ac8c-07e986bfa75a  
653              N  b4916a3a-3eea-416a-b567-1192a29955a2  
1864             Y  7aa13c15-0660-4e88-860a-7f2f0b378c82  
1407             Y  9ec0468b-9fd6-4aa4-90c1-4d86b0df5a0f  
427              N  07f5c424-4295-4620-8a97-a31d73d3e808  
74               Y  a69b26b1-45e4-4c61-8a00-866ec060ca52  

[10 rows x 52 columns]

In [3]:
break_data.shape

(2670, 52)

In [4]:
break_data.columns

Index(['X', 'Y', 'OBJECTID', 'WATBREAKINCIDENTID', 'INCIDENT_DATE',
       'BREAK_TYPE', 'ROAD_CLOSED', 'SIDEWALK_CLOSED', 'HOUR_IMPACTED',
       'UNITS_IMPACTED', 'CW_SERVICE_REQUEST', 'STATUS', 'STATUS_DATE',
       'WORKORDER', 'RETURN_TO_NORMAL', 'BREAK_NATURE', 'BREAK_APPARENT_CAUSE',
       'REPAIR_TYPE', 'NEW_SECTION_LENGTH', 'MAINTENANCE_DESC',
       'VALVES_CLOSED', 'VALVES_OPENED', 'HYDRANTS_CALLED_OUT',
       'HYDRANTS_CALLED_BACK_IN', 'POSITIVE_PRESSURE_MAINTANED',
       'AIR_GAP_MAINTANED', 'DISINFECTED', 'MECHANICAL_REMOVAL',
       'FLUSHING_EXCAVATION', 'HIGHER_VELOCITY_FLUSHING', 'ANODE_INSTALLED',
       'BREAK_CATEGORIZATION', 'BACTERIA_TESTING_DATE',
       'HEALTH_DEPT_NOTIFICATION', 'MOECC_SAC_NOTIFICATION',
       'SAC_REFERENCE_NO', 'LOCAL_MOE_OFFICE', 'BWA_DWA', 'BWA_DWA_DECLARED',
       'PROCEEDURES_FOLLOWED', 'RECORD_CHANGE_REQD', 'ROADSEGMENTID',
       'CIVIC_NUMBER', 'STREET', 'ASSETID', 'ASSET_DEPTH', 'FROST_DEPTH',
       'ASSET_SIZE', 'ASSET_YEAR_I

Column descriptions from :

- Wat Break Incident ID
- Incident date
- Type of Asset Broken
- Does the road need to be closed?
- Does the sidewalk need to be closed?
- Estimated Hours for Repair
- Estimated Number of Units Impacted
- CW Service Request Number
- Current status of the break
- Status last updated date
- CW Workorder #
- Date operations was returned to normal service
- Nature of Break
- Apparent cause of break
- Repair Type
- New Section Length (m)
- Type of Planned Maintenance
- List Valves Closed
- List Hydrants Called Out
- List Hydrants Called Back In
- Positive Pressure Maintained?
- Air Gap Maintained thr Repair Process?
- Pipe and Press Parts Disinfected?
- Mechanical removal of contaminants?
- Flushing into the excavation?
- Higher velocity flushing after repairs?
- Anode Installed?
- Categorization of the Break
- Bacteria Testing Date Taken
- Health Dept Notification
- MOECC/SAC Notification
- Local MOE Office Notification
- BWA/DWA declared
- IF BWA / DWA are issued, Date/Time declared
- Where proceedures Followed?
- Was a record change required?
- Road Segment ID
- Closest Civic Number
- Street
- Related Asset ID
- Related Asset Depth (m)
- Depth of Frost (m)
- Asset Size (cm)
- Year Asset Installed
- Asset Material
- Asset Exists

In [5]:
break_data.describe()

X            Y      OBJECTID  WATBREAKINCIDENTID  \
count  2670.000000  2670.000000   2670.000000         2670.000000   
mean    -80.477581    43.442219  10871.607865        13656.689139   
std       0.031289     0.017289   6519.919953        42277.308985   
min     -80.573004    43.380008      1.000000            1.000000   
25%     -80.504334    43.429881   8652.250000          670.250000   
50%     -80.472111    43.441523   9319.500000         1342.500000   
75%     -80.453397    43.452611   9986.750000         2011.750000   
max     -80.294712    43.538106  48321.000000       197291.000000   

       CW_SERVICE_REQUEST      WORKORDER  NEW_SECTION_LENGTH  \
count        1.900000e+01    1496.000000            8.000000   
mean        -6.691076e+07  140914.652406            5.885000   
std          2.938190e+08   89207.185563           12.208976   
min         -1.280223e+09     534.000000            0.500000   
25%          2.196830e+05   69223.250000            0.925000   
50%          2.394500e+05  130919.000000            1.625000   
75%          2.397180e+05  200446.000000            2.407500   
max          5.033240e+06  329537.000000           36.000000   

       MAINTENANCE_DESC  VALVES_OPENED  HYDRANTS_CALLED_OUT  \
count               0.0            0.0                  1.0   
mean                NaN            NaN                353.0   
std                 NaN            NaN                  NaN   
min                 NaN            NaN                353.0   
25%                 NaN            NaN                353.0   
50%                 NaN            NaN                353.0   
75%                 NaN            NaN                353.0   
max                 NaN            NaN                353.0   

       HYDRANTS_CALLED_BACK_IN  HEALTH_DEPT_NOTIFICATION  SAC_REFERENCE_NO  \
count                      1.0                       0.0               0.0   
mean                     353.0                       NaN               NaN   
std                        NaN                       NaN               NaN   
min                      353.0                       NaN               NaN   
25%                      353.0                       NaN               NaN   
50%                      353.0                       NaN               NaN   
75%                      353.0                       NaN               NaN   
max                      353.0                       NaN               NaN   

       ROADSEGMENTID       ASSETID  ASSET_DEPTH  FROST_DEPTH   ASSET_SIZE  \
count    2670.000000  2.670000e+03    17.000000    42.000000  2549.000000   
mean    20597.352809  8.114945e+04     1.747059     0.426190   183.763829   
std     64996.511242  4.019611e+05     0.292176     0.450777    88.249107   
min         0.000000  0.000000e+00     1.270000     0.000000     0.000000   
25%      8682.000000  1.728250e+04     1.524000     0.000000   150.000000   
50%     11171.000000  3.353000e+04     1.830000     0.500000   150.000000   
75%     15884.000000  7.572825e+04     1.830000     0.750000   200.000000   
max    606029.000000  4.769517e+06     2.438000     2.000000  1200.000000   

       ASSET_YEAR_INSTALLED  
count           2544.000000  
mean            1968.507469  
std               21.438055  
min             1889.000000  
25%             1957.000000  
50%             1964.000000  
75%             1972.000000  
max             2018.000000

In [6]:
break_data.isna().sum()

X                                 0
Y                                 0
OBJECTID                          0
WATBREAKINCIDENTID                0
INCIDENT_DATE                     0
BREAK_TYPE                        0
ROAD_CLOSED                       0
SIDEWALK_CLOSED                   0
HOUR_IMPACTED                     0
UNITS_IMPACTED                 2472
CW_SERVICE_REQUEST             2651
STATUS                            0
STATUS_DATE                     167
WORKORDER                      1174
RETURN_TO_NORMAL               2621
BREAK_NATURE                     89
BREAK_APPARENT_CAUSE            127
REPAIR_TYPE                    2567
NEW_SECTION_LENGTH             2662
MAINTENANCE_DESC               2670
VALVES_CLOSED                  2669
VALVES_OPENED                  2670
HYDRANTS_CALLED_OUT            2669
HYDRANTS_CALLED_BACK_IN        2669
POSITIVE_PRESSURE_MAINTANED       0
AIR_GAP_MAINTANED                 0
DISINFECTED                       0
MECHANICAL_REMOVAL          

In [7]:
pd.set_option("display.max_columns", None)

break_data.sample(10)

X          Y  OBJECTID  WATBREAKINCIDENTID  \
1452 -80.527825  43.440396      9437                 892   
584  -80.524333  43.438804      8569                1347   
1935 -80.444354  43.442146      9920                1584   
2354 -80.436134  43.433234     10339                2311   
186  -80.444645  43.462990      8171                1544   
86   -80.491983  43.450301      8071                  61   
1040 -80.452365  43.450516      9025                2066   
1799 -80.492607  43.477228      9784                 882   
859  -80.431398  43.434300      8844                 709   
1365 -80.448065  43.445243      9350                 753   

               INCIDENT_DATE BREAK_TYPE ROAD_CLOSED SIDEWALK_CLOSED  \
1452  1999/07/05 00:00:00+00       MAIN        Open            Open   
584   2011/12/15 15:38:28+00       MAIN        Open            Open   
1935  2011/01/21 00:00:00+00       MAIN        Open            Open   
2354  2017/12/22 00:00:00+00       MAIN        Open            Open   
186   2010/12/10 00:00:00+00       MAIN        Open            Open   
86    2003/01/23 00:00:00+00       MAIN        Open            Open   
1040  2015/02/26 00:00:00+00       MAIN        Open            Open   
1799  2004/02/25 00:00:00+00       MAIN        Open            Open   
859   2000/04/14 00:00:00+00       MAIN        Open            Open   
1365  2005/03/14 00:00:00+00       MAIN        Open            Open   

     HOUR_IMPACTED UNITS_IMPACTED  CW_SERVICE_REQUEST            STATUS  \
1452    8-12 hours            NaN                 NaN  REPAIR COMPLETED   
584     8-12 hours            NaN                 NaN  REPAIR COMPLETED   
1935    8-12 hours            NaN                 NaN  REPAIR COMPLETED   
2354    8-12 hours            NaN                 NaN  REPAIR COMPLETED   
186     8-12 hours            NaN                 NaN  REPAIR COMPLETED   
86      8-12 hours            NaN                 NaN  REPAIR COMPLETED   
1040    8-12 hours            NaN                 NaN  REPAIR COMPLETED   
1799    8-12 hours            NaN                 NaN  REPAIR COMPLETED   
859     8-12 hours            NaN                 NaN  REPAIR COMPLETED   
1365    8-12 hours            NaN                 NaN  REPAIR COMPLETED   

                 STATUS_DATE  WORKORDER RETURN_TO_NORMAL     BREAK_NATURE  \
1452  2011/12/15 15:38:27+00        NaN              NaN          UNKNOWN   
584   2011/12/15 15:38:28+00     3595.0              NaN          UNKNOWN   
1935  2011/12/15 15:38:26+00    39681.0              NaN  CIRCUMFERENTIAL   
2354  2018/11/27 00:00:00+00   200331.0              NaN          UNKNOWN   
186   2011/12/15 15:38:27+00    38239.0              NaN  CIRCUMFERENTIAL   
86    2011/12/15 15:38:26+00        NaN              NaN          UNKNOWN   
1040  2015/04/15 00:00:00+00   127177.0              NaN          UNKNOWN   
1799  2011/12/15 15:38:27+00        NaN              NaN          UNKNOWN   
859   2011/12/15 15:38:27+00        NaN              NaN          UNKNOWN   
1365  2011/12/15 15:38:27+00        NaN              NaN          UNKNOWN   

     BREAK_APPARENT_CAUSE REPAIR_TYPE  NEW_SECTION_LENGTH  MAINTENANCE_DESC  \
1452                OTHER         NaN                 NaN               NaN   
584                 OTHER         NaN                 NaN               NaN   
1935             PRESSURE         NaN                 NaN               NaN   
2354                OTHER         NaN                 NaN               NaN   
186                   AGE         NaN                 NaN               NaN   
86                  OTHER         NaN                 NaN               NaN   
1040                OTHER         NaN                 NaN               NaN   
1799                OTHER         NaN                 NaN               NaN   
859                 OTHER         NaN                 NaN               NaN   
1365                OTHER         NaN                 NaN               NaN   

     VALVES_CLOSED  VALVES_

In [8]:
break_data[~break_data['UNITS_IMPACTED'].isna()]

X          Y  OBJECTID  WATBREAKINCIDENTID  \
0    -80.484005  43.462939         1                2252   
2472 -80.454339  43.381667     10881              136233   
2474 -80.520814  43.441519     12161              185060   
2475 -80.452472  43.466348     12481              185178   
2476 -80.434568  43.432085     12801              185295   
...         ...        ...       ...                 ...   
2665 -80.510704  43.421611     47041              144634   
2666 -80.511861  43.405537     47361              144654   
2667 -80.431720  43.431982     47681              144674   
2668 -80.530689  43.441695     48001              144694   
2669 -80.461043  43.439032     48321              144714   

               INCIDENT_DATE BREAK_TYPE       ROAD_CLOSED SIDEWALK_CLOSED  \
0     2017/12/01 15:15:00+00       MAIN  Partially Closed            Open   
2472  2019/10/01 15:23:00+00    SERVICE              Open            Open   
2474  2019/11/06 13:24:53+00       MAIN            Closed            Open   
2475  2019/11/08 20:47:10+00       MAIN            Closed            Open   
2476  2019/11/15 11:29:46+00       MAIN            Closed            Open   
...                      ...        ...               ...             ...   
2665  2022/01/15 15:37:12+00       MAIN              Open            Open   
2666  2022/01/16 21:30:50+00       MAIN            Closed            Open   
2667  2022/01/17 20:58:57+00       MAIN            Closed            Open   
2668  2022/01/18 13:30:27+00       MAIN            Closed            Open   
2669  2022/01/21 06:51:36+00       MAIN            Closed            Open   

     HOUR_IMPACTED UNITS_IMPACTED  CW_SERVICE_REQUEST               STATUS  \
0      12-16 hours             47        1.234560e+05     REPAIR COMPLETED   
2472   20-24 hours              0                 NaN     REPAIR COMPLETED   
2474     4-8 hours           0-50        5.033240e+06     REPAIR COMPLETED   
2475    8-12 hours           0-50        2.158950e+05     REPAIR COMPLETED   
2476     4-8 hours         50-100       -1.280223e+09     REPAIR COMPLETED   
...            ...            ...                 ...                  ...   
2665     4-8 hours           0-50                 NaN     REPAIR COMPLETED   
2666     4-8 hours           0-50                 NaN     REPAIR COMPLETED   
2667     4-8 hours           0-50                 NaN     REPAIR COMPLETED   
2668     4-8 hours           0-50                 NaN  UNDER INVESTIGATION   
2669     4-8 hours           0-50                 NaN  UNDER INVESTIGATION   

                 STATUS_DATE  WORKORDER        RETURN_TO_NORMAL  \
0     2017/12/02 00:00:00+00   177898.0  2017/01/31 23:30:00+00   
2472                     NaN        NaN                     NaN   
2474                     NaN   260276.0                     NaN   
2475                     NaN   260710.0                     NaN   
2476                     NaN   261125.0                     NaN   
...                      ...        ...                     ...   
2665  2022/01/16 09:58:35+00   329235.0  2022/01/16 05:30:46+00   
2666  2022/01/17 17:28:55+00   329238.0  2022/01/17 22:30:51+00   
2667  2022/01/18 14:05:58+00   329315.0                     NaN   
2668  2022/01/18 13:30:27+00   329367.0                     NaN   
2669  2022/01/21 06:51:36+00   329537.0                     NaN   

                                   BREAK_NATURE BREAK_APPARENT_CAUSE  \
0                   CORROSION AND FITTING/JOINT                  AGE   
2472                                      OTHER              UNKNOWN   
2474                                        NaN                  NaN   
2475                                        NaN                  NaN   
2476                            CIRCUMFERENTIAL                  AGE   
...                                         ...                  ...   
2665                            CIRCUMFERENTIAL              UNKNOWN   
2666  CORROSION - ROBAR SADDLE CORRODED AT SEAM   

In [9]:
break_data['UNITS_IMPACTED'] = break_data['UNITS_IMPACTED'].fillna(0)

In [10]:
break_data = break_data.rename(columns={'X': 'LONGITUDE', 'Y': 'LATITUDE'})

In [11]:
break_data['INCIDENT_DATE'] = pd.to_datetime(break_data['INCIDENT_DATE'])

In [12]:
break_data.head()

LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
0 -80.484005  43.462939         1                2252   
1 -80.515075  43.422742      7874                1311   
2 -80.439811  43.445067      7875                1325   
3 -80.510859  43.426478      7876                1328   
4 -80.457520  43.443201      7877                1308   

              INCIDENT_DATE BREAK_TYPE       ROAD_CLOSED SIDEWALK_CLOSED  \
0 2017-12-01 15:15:00+00:00       MAIN  Partially Closed            Open   
1 2001-03-26 00:00:00+00:00    SERVICE              Open            Open   
2 2006-09-06 00:00:00+00:00    SERVICE              Open            Open   
3 2006-09-11 00:00:00+00:00    SERVICE              Open            Open   
4 2000-01-27 00:00:00+00:00    SERVICE              Open            Open   

  HOUR_IMPACTED UNITS_IMPACTED  CW_SERVICE_REQUEST            STATUS  \
0   12-16 hours             47            123456.0  REPAIR COMPLETED   
1    8-12 hours              0                 NaN  REPAIR COMPLETED   
2    8-12 hours              0                 NaN  REPAIR COMPLETED   
3    8-12 hours              0                 NaN  REPAIR COMPLETED   
4    8-12 hours              0                 NaN  REPAIR COMPLETED   

              STATUS_DATE  WORKORDER        RETURN_TO_NORMAL  \
0  2017/12/02 00:00:00+00   177898.0  2017/01/31 23:30:00+00   
1  2011/12/15 15:38:28+00        NaN                     NaN   
2  2011/12/15 15:38:28+00   162230.0                     NaN   
3  2011/12/15 15:38:28+00        NaN                     NaN   
4  2011/12/15 15:38:28+00        NaN                     NaN   

                  BREAK_NATURE BREAK_APPARENT_CAUSE REPAIR_TYPE  \
0  CORROSION AND FITTING/JOINT                  AGE      CUTOUT   
1                      UNKNOWN                OTHER         NaN   
2                      UNKNOWN                OTHER         NaN   
3                      UNKNOWN                OTHER         NaN   
4                      UNKNOWN                OTHER         NaN   

   NEW_SECTION_LENGTH  MAINTENANCE_DESC  \
0                 2.0               NaN   
1                 NaN               NaN   
2                 NaN               NaN   
3                 NaN               NaN   
4                 NaN               NaN   

                                       VALVES_CLOSED  VALVES_OPENED  \
0  6" 93563, 94817, 93902,1139; 18"- 1264, 92862,...            NaN   
1                                                NaN            NaN   
2                                                NaN            NaN   
3                                                NaN            NaN   
4                                                NaN            NaN   

   HYDRANTS_CALLED_OUT  HYDRANTS_CALLED_BACK_IN POSITIVE_PRESSURE_MAINTANED  \
0                353.0                    353.0                           N   
1                  NaN                      NaN                           Y   
2                  NaN                      NaN                           Y   
3                  NaN                      NaN                           Y   
4                  NaN                      NaN                           Y   

  AIR_GAP_MAINTANED DISINFECTED MECHANICAL_REMOVAL FLUSHING_EXCAVATION  \
0                 N           Y                  N                   N   
1                 Y           Y                  Y                   Y   
2                 Y           Y                  Y                   Y   
3                 Y           Y                  Y                   Y   
4                 Y           Y                  Y                   Y   

  HIGHER_VELOCITY_FLUSHING ANODE_INSTALLED BREAK_CATEGORIZATION  \
0                        Y               Y           CATEGORY 2   
1                        Y               Y           CATEGORY 1   
2                        Y               Y           CATEGORY 1   
3                        Y               Y           CATEGORY 1   
4                        Y               

There are a lot of columns in this data giving various sorts of information. I feel like only some are useful and a lot of them I'll have to sift through and see if they're really even relevant. Columns `CW_SERVICE_REQUEST`, `WORKORDER`, `HEALTH_DEPT_NOTIFICATION`, `SAC_REFERENCE_NO`, `LOCAL_MOE_OFFICE`, BWA_DWA`, `BWA_DWA_DECLARED` all seem like they won't contribute much information so we'll exclude those, and possible more as I go through the data.

In [13]:
break_data = break_data.drop(['CW_SERVICE_REQUEST', 'WORKORDER', 'HEALTH_DEPT_NOTIFICATION',
                              'SAC_REFERENCE_NO', 'LOCAL_MOE_OFFICE', 'BWA_DWA', 'BWA_DWA_DECLARED',
                              'PROCEEDURES_FOLLOWED', 'RECORD_CHANGE_REQD', 'MOECC_SAC_NOTIFICATION',
                              'BACTERIA_TESTING_DATE'], axis=1)

In [14]:
print(break_data.shape)
break_data.sample(5)

(2670, 41)


LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
1705 -80.488568  43.469833      9690                 132   
500  -80.460124  43.431056      8485                1986   
968  -80.509190  43.433803      8953                 271   
1957 -80.463369  43.427808      9942                1598   
913  -80.441351  43.456089      8898                 675   

                 INCIDENT_DATE BREAK_TYPE ROAD_CLOSED SIDEWALK_CLOSED  \
1705 2003-03-21 00:00:00+00:00       MAIN        Open            Open   
500  2014-05-14 00:00:00+00:00       MAIN        Open            Open   
968  2002-12-02 00:00:00+00:00       MAIN        Open            Open   
1957 2011-02-11 00:00:00+00:00       MAIN        Open            Open   
913  2002-02-13 00:00:00+00:00       MAIN        Open            Open   

     HOUR_IMPACTED UNITS_IMPACTED            STATUS             STATUS_DATE  \
1705    8-12 hours              0  REPAIR COMPLETED  2011/12/15 15:38:26+00   
500     8-12 hours              0  REPAIR COMPLETED  2015/01/14 00:00:00+00   
968     8-12 hours              0  REPAIR COMPLETED  2011/12/15 15:38:26+00   
1957    8-12 hours              0  REPAIR COMPLETED  2011/12/15 15:38:26+00   
913     8-12 hours              0  REPAIR COMPLETED  2011/12/15 15:38:27+00   

     RETURN_TO_NORMAL   BREAK_NATURE BREAK_APPARENT_CAUSE REPAIR_TYPE  \
1705              NaN        UNKNOWN                OTHER         NaN   
500               NaN        UNKNOWN                OTHER         NaN   
968               NaN        UNKNOWN                OTHER         NaN   
1957              NaN  FITTING/JOINT          COMBINATION         NaN   
913               NaN        UNKNOWN                OTHER         NaN   

      NEW_SECTION_LENGTH  MAINTENANCE_DESC VALVES_CLOSED  VALVES_OPENED  \
1705                 NaN               NaN           NaN            NaN   
500                  NaN               NaN           NaN            NaN   
968                  NaN               NaN           NaN            NaN   
1957                 NaN               NaN           NaN            NaN   
913                  NaN               NaN           NaN            NaN   

      HYDRANTS_CALLED_OUT  HYDRANTS_CALLED_BACK_IN  \
1705                  NaN                      NaN   
500                   NaN                      NaN   
968                   NaN                      NaN   
1957                  NaN                      NaN   
913                   NaN                      NaN   

     POSITIVE_PRESSURE_MAINTANED AIR_GAP_MAINTANED DISINFECTED  \
1705                           Y                 Y           Y   
500                            Y                 Y           Y   
968                            Y                 Y           Y   
1957                           Y                 Y           Y   
913                            Y                 Y           Y   

     MECHANICAL_REMOVAL FLUSHING_EXCAVATION HIGHER_VELOCITY_FLUSHING  \
1705                  Y                   Y                        Y   
500                   Y                   Y                        Y   
968                   Y                   Y                        Y   
1957                  Y                   Y                        Y   
913                   Y                   Y                        Y   

     ANODE_INSTALLED BREAK_CATEGORIZATION  ROADSEGMENTID CIVIC_NUMBER  \
1705               Y           CATEGORY 1          10442          332   
500                Y           CATEGORY 1          70277            8   
968                Y           CATEGORY 1          20924         1230   
1957               Y           CATEGORY 1           9411          132   
913                Y           CATEGORY 1          23231           58   

             STREET  ASSETID  ASSET_DEPTH  FROST_DEPTH  ASSET_SIZE  \
1705       EDWIN ST    11600          NaN          NaN       150.0   
500      SHELLEY PL    34770          NaN          NaN       100.0   
968     QUEENS BLVD    31750         

In [15]:
break_data.isnull().sum()

LONGITUDE                         0
LATITUDE                          0
OBJECTID                          0
WATBREAKINCIDENTID                0
INCIDENT_DATE                     0
BREAK_TYPE                        0
ROAD_CLOSED                       0
SIDEWALK_CLOSED                   0
HOUR_IMPACTED                     0
UNITS_IMPACTED                    0
STATUS                            0
STATUS_DATE                     167
RETURN_TO_NORMAL               2621
BREAK_NATURE                     89
BREAK_APPARENT_CAUSE            127
REPAIR_TYPE                    2567
NEW_SECTION_LENGTH             2662
MAINTENANCE_DESC               2670
VALVES_CLOSED                  2669
VALVES_OPENED                  2670
HYDRANTS_CALLED_OUT            2669
HYDRANTS_CALLED_BACK_IN        2669
POSITIVE_PRESSURE_MAINTANED       0
AIR_GAP_MAINTANED                 0
DISINFECTED                       0
MECHANICAL_REMOVAL                0
FLUSHING_EXCAVATION               0
HIGHER_VELOCITY_FLUSHING    

In [16]:
break_data[~break_data['RETURN_TO_NORMAL'].isna()]

LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
0    -80.484005  43.462939         1                2252   
2508 -80.454120  43.462109     20162              193136   
2513 -80.462563  43.425671     21441              193316   
2521 -80.486162  43.458394     23041              193577   
2530 -80.534138  43.441349     24669              138494   
2533 -80.477244  43.444133     24961              138596   
2543 -80.484132  43.456092     27201              138974   
2552 -80.499831  43.462908     29761              139474   
2556 -80.436444  43.433055     31361              139654   
2561 -80.431032  43.431434     32002              139934   
2562 -80.510249  43.423448     32321              139954   
2563 -80.450730  43.455416     32641              139974   
2566 -80.448923  43.443261     32961              139994   
2569 -80.474602  43.459899     33282              140034   
2570 -80.526126  43.445705     33283              140036   
2571 -80.457848  43.460685     33284              140054   
2573 -80.474592  43.459890     33921              140184   
2579 -80.489654  43.467870     33927              140298   
2580 -80.452660  43.429538     34241              140314   
2587 -80.463707  43.437574     35201              140494   
2589 -80.464781  43.484149     35203              140614   
2591 -80.512528  43.432558     35841              140654   
2592 -80.461296  43.439222     35842              140656   
2593 -80.426523  43.434735     36161              140674   
2600 -80.491765  43.435026     36801              140736   
2603 -80.471460  43.456771     37123              140776   
2607 -80.457940  43.452449     37442              140796   
2612 -80.452083  43.451289     38083              141156   
2613 -80.472193  43.453985     38402              141274   
2616 -80.515381  43.433265     39042              141414   
2619 -80.452885  43.427967     39363              141594   
2621 -80.441137  43.466287     40001              142194   
2622 -80.473101  43.458588     40323              142258   
2623 -80.428474  43.428408     40642              142413   
2633 -80.504583  43.429187     42561              142753   
2637 -80.431022  43.428065     43521              143074   
2638 -80.514634  43.434060     43522              143114   
2639 -80.439180  43.450646     43841              143374   
2642 -80.467544  43.428456     44161              144174   
2645 -80.453961  43.461882     44482              144234   
2647 -80.525628  43.421387     44484              144274   
2651 -80.528348  43.447502     45122              144374   
2653 -80.449875  43.445184     45124              144394   
2654 -80.509771  43.424092     45125              144414   
2657 -80.448114  43.442294     45761              144474   
2658 -80.469004  43.462795     46081              144494   
2659 -80.526013  43.425227     46082              144514   
2665 -80.510704  43.421611     47041              144634   
2666 -80.511861  43.405537     47361              144654   

                 INCIDENT_DATE BREAK_TYPE        ROAD_CLOSED  \
0    2017-12-01 15:15:00+00:00       MAIN   Partially Closed   
2508 2020-01-21 07:58:25+00:00       MAIN             Closed   
2513 2020-02-06 16:24:47+00:00       MAIN               Open   
2521 2020-02-29 04:30:00+00:00       MAIN             Closed   
2530 2020-06-17 08:01:33+00:00    SERVICE             Closed   
2533 2020-06-25 19:54:53+00:00       MAIN             Closed   
2543 2020-08-03 00:44:49+00:00       MAIN             Closed   
2552 2020-10-31 10:58:26+00:00       MAIN             Closed   
2556 2020-11-20 13:40:04+00:00       MAIN   Partially Closed   
2561 2020-12-19 06:41:13+00:00       MAIN             Closed   
2562 2020-12-21 14:52:11+00:00       MAIN             Closed   
2563 2020-12-23 08:32:32+00:00       MAIN             Closed   
2566 2020-12-24 10:06:01+00:00       MAIN             Closed   
2569 2020-12-31 03:31:05+00:00       MAIN   Partially Closed   
2570 2020-12-31 09:57:42+00:00       MAIN    

In [17]:
break_data = break_data.drop(['STATUS_DATE', 'MAINTENANCE_DESC', 'VALVES_CLOSED', 'VALVES_OPENED',
                              'HYDRANTS_CALLED_OUT', 'HYDRANTS_CALLED_BACK_IN'], axis=1)

In [18]:
break_data.head()

LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
0 -80.484005  43.462939         1                2252   
1 -80.515075  43.422742      7874                1311   
2 -80.439811  43.445067      7875                1325   
3 -80.510859  43.426478      7876                1328   
4 -80.457520  43.443201      7877                1308   

              INCIDENT_DATE BREAK_TYPE       ROAD_CLOSED SIDEWALK_CLOSED  \
0 2017-12-01 15:15:00+00:00       MAIN  Partially Closed            Open   
1 2001-03-26 00:00:00+00:00    SERVICE              Open            Open   
2 2006-09-06 00:00:00+00:00    SERVICE              Open            Open   
3 2006-09-11 00:00:00+00:00    SERVICE              Open            Open   
4 2000-01-27 00:00:00+00:00    SERVICE              Open            Open   

  HOUR_IMPACTED UNITS_IMPACTED            STATUS        RETURN_TO_NORMAL  \
0   12-16 hours             47  REPAIR COMPLETED  2017/01/31 23:30:00+00   
1    8-12 hours              0  REPAIR COMPLETED                     NaN   
2    8-12 hours              0  REPAIR COMPLETED                     NaN   
3    8-12 hours              0  REPAIR COMPLETED                     NaN   
4    8-12 hours              0  REPAIR COMPLETED                     NaN   

                  BREAK_NATURE BREAK_APPARENT_CAUSE REPAIR_TYPE  \
0  CORROSION AND FITTING/JOINT                  AGE      CUTOUT   
1                      UNKNOWN                OTHER         NaN   
2                      UNKNOWN                OTHER         NaN   
3                      UNKNOWN                OTHER         NaN   
4                      UNKNOWN                OTHER         NaN   

   NEW_SECTION_LENGTH POSITIVE_PRESSURE_MAINTANED AIR_GAP_MAINTANED  \
0                 2.0                           N                 N   
1                 NaN                           Y                 Y   
2                 NaN                           Y                 Y   
3                 NaN                           Y                 Y   
4                 NaN                           Y                 Y   

  DISINFECTED MECHANICAL_REMOVAL FLUSHING_EXCAVATION HIGHER_VELOCITY_FLUSHING  \
0           Y                  N                   N                        Y   
1           Y                  Y                   Y                        Y   
2           Y                  Y                   Y                        Y   
3           Y                  Y                   Y                        Y   
4           Y                  Y                   Y                        Y   

  ANODE_INSTALLED BREAK_CATEGORIZATION  ROADSEGMENTID CIVIC_NUMBER  \
0               Y           CATEGORY 2           6961          125   
1               Y           CATEGORY 1           6116           76   
2               Y           CATEGORY 1          13207           47   
3               Y           CATEGORY 1           6498          382   
4               Y           CATEGORY 1          22846          224   

            STREET  ASSETID  ASSET_DEPTH  FROST_DEPTH  ASSET_SIZE  \
0   LANCASTER ST W   134292          1.6          0.3       450.0   
1  CLOVERDALE CRES  4101323          NaN          NaN        13.0   
2        WREN CRES  4099987          NaN          NaN        25.0   
3    GREENBROOK DR  4642530          NaN          NaN        25.0   
4    MONTGOMERY RD  4100648          NaN          NaN        25.0   

   ASSET_YEAR_INSTALLED ASSET_MATERIAL ASSET_EXISTS  \
0                1937.0             CI            Y   
1                1965.0            XXX            Y   
2                1967.0            XXX            Y   
3                1964.0            PVC            Y   
4                1967.0            XXX            Y   

                               GLOBALID  
0  3521d297-1a2e-4e7b-a071-fc53ed87e965  
1  72445d62-16a8-43c1-9733-56b06015b077  
2  3bdc8931-31c0-4090-a07a-a6847781dd97  
3  f75ad0b1-5b2a-4125-8ad5-2b9a037debd7  
4  5a3c5d03-0899-4899-95e7-278bc5cbb682

In [19]:
break_data = break_data.drop(['RETURN_TO_NORMAL', 'REPAIR_TYPE', 'NEW_SECTION_LENGTH', 'CIVIC_NUMBER'], axis=1)

In [20]:
break_data.shape

(2670, 31)

We've cut down our features from 57 to 31 so far. I wish there were less values missing from `ASSET_DEPTH` and `FROST_DEPTH` because I think they would be valuable for our analysis, but unfortunately I will have to get rid of them as well since almost all of the values are missing.

From the studies that I've read, none of the researchers used features related to asset depth or frost depth so my hypothesis about them providing valuable info could be wrong anyway and just a bad judgment call.

In [21]:
break_data = break_data.drop(['ASSET_DEPTH', 'FROST_DEPTH'], axis=1)

In [22]:
break_data.isna().sum()

LONGITUDE                        0
LATITUDE                         0
OBJECTID                         0
WATBREAKINCIDENTID               0
INCIDENT_DATE                    0
BREAK_TYPE                       0
ROAD_CLOSED                      0
SIDEWALK_CLOSED                  0
HOUR_IMPACTED                    0
UNITS_IMPACTED                   0
STATUS                           0
BREAK_NATURE                    89
BREAK_APPARENT_CAUSE           127
POSITIVE_PRESSURE_MAINTANED      0
AIR_GAP_MAINTANED                0
DISINFECTED                      0
MECHANICAL_REMOVAL               0
FLUSHING_EXCAVATION              0
HIGHER_VELOCITY_FLUSHING         0
ANODE_INSTALLED                  0
BREAK_CATEGORIZATION           105
ROADSEGMENTID                    0
STREET                          16
ASSETID                          0
ASSET_SIZE                     121
ASSET_YEAR_INSTALLED           126
ASSET_MATERIAL                 121
ASSET_EXISTS                     0
GLOBALID            

In [23]:
break_data[break_data['BREAK_APPARENT_CAUSE'].isna()]

LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
2473 -80.448558  43.459837     11841              184820   
2474 -80.520814  43.441519     12161              185060   
2475 -80.452472  43.466348     12481              185178   
2477 -80.445655  43.451450     13121              185336   
2478 -80.451158  43.437402     13123              185365   
...         ...        ...       ...                 ...   
2661 -80.433710  43.434485     46402              144536   
2663 -80.530123  43.424289     46721              144574   
2664 -80.513145  43.425433     46722              144576   
2668 -80.530689  43.441695     48001              144694   
2669 -80.461043  43.439032     48321              144714   

                 INCIDENT_DATE BREAK_TYPE ROAD_CLOSED SIDEWALK_CLOSED  \
2473 2019-11-01 17:11:41+00:00       MAIN      Closed            Open   
2474 2019-11-06 13:24:53+00:00       MAIN      Closed            Open   
2475 2019-11-08 20:47:10+00:00       MAIN      Closed            Open   
2477 2019-11-18 10:06:59+00:00       MAIN      Closed            Open   
2478 2019-11-19 10:53:51+00:00       MAIN      Closed          Closed   
...                        ...        ...         ...             ...   
2661 2022-01-05 10:29:04+00:00       MAIN        Open            Open   
2663 2022-01-08 09:38:07+00:00       MAIN        Open            Open   
2664 2022-01-08 13:31:58+00:00       MAIN        Open            Open   
2668 2022-01-18 13:30:27+00:00       MAIN      Closed            Open   
2669 2022-01-21 06:51:36+00:00       MAIN      Closed            Open   

     HOUR_IMPACTED UNITS_IMPACTED               STATUS BREAK_NATURE  \
2473   20-24 hours              0     REPAIR COMPLETED          NaN   
2474     4-8 hours           0-50     REPAIR COMPLETED          NaN   
2475    8-12 hours           0-50     REPAIR COMPLETED          NaN   
2477    8-12 hours           0-50     REPAIR COMPLETED          NaN   
2478    8-12 hours           0-50     REPAIR COMPLETED          NaN   
...            ...            ...                  ...          ...   
2661     4-8 hours           0-50     REPAIR COMPLETED          NaN   
2663     4-8 hours           0-50     REPAIR COMPLETED          NaN   
2664     4-8 hours           0-50     REPAIR COMPLETED          NaN   
2668     4-8 hours           0-50  UNDER INVESTIGATION          NaN   
2669     4-8 hours           0-50  UNDER INVESTIGATION          NaN   

     BREAK_APPARENT_CAUSE POSITIVE_PRESSURE_MAINTANED AIR_GAP_MAINTANED  \
2473                  NaN                           Y                 Y   
2474                  NaN                           Y                 Y   
2475                  NaN                           Y                 Y   
2477                  NaN                           Y                 Y   
2478                  NaN                           Y                 Y   
...                   ...                         ...               ...   
2661                  NaN                           Y                 Y   
2663                  NaN                           Y                 Y   
2664                  NaN                           Y                 Y   
2668                  NaN                           Y                 Y   
2669                  NaN                           Y                 Y   

     DISINFECTED MECHANICAL_REMOVAL FLUSHING_EXCAVATION  \
2473           Y                  Y                   Y   
2474           Y                  Y                   Y   
2475           Y                  Y                   Y   
2477           Y                  Y                   Y   
2478           Y                  Y                   Y   
...          ...                ...                 ...   
2661           Y                  Y                   Y   
2663           Y                  Y                   Y   
2664           Y                  Y                   Y   
2668           Y                  Y                   Y   
2669 

In [24]:
break_data.BREAK_APPARENT_CAUSE.unique()

array(['AGE', 'OTHER', 'COMBINATION', 'CORROSION', 'SOILS', 'UNKNOWN',
       'PRESSURE', 'FAULTY INSTALL', nan], dtype=object)

In [25]:
break_data.BREAK_APPARENT_CAUSE.fillna('UNKNOWN', inplace=True)

In [26]:
break_data.BREAK_APPARENT_CAUSE.unique()

array(['AGE', 'OTHER', 'COMBINATION', 'CORROSION', 'SOILS', 'UNKNOWN',
       'PRESSURE', 'FAULTY INSTALL'], dtype=object)

In [27]:
break_data.isna().sum()

LONGITUDE                        0
LATITUDE                         0
OBJECTID                         0
WATBREAKINCIDENTID               0
INCIDENT_DATE                    0
BREAK_TYPE                       0
ROAD_CLOSED                      0
SIDEWALK_CLOSED                  0
HOUR_IMPACTED                    0
UNITS_IMPACTED                   0
STATUS                           0
BREAK_NATURE                    89
BREAK_APPARENT_CAUSE             0
POSITIVE_PRESSURE_MAINTANED      0
AIR_GAP_MAINTANED                0
DISINFECTED                      0
MECHANICAL_REMOVAL               0
FLUSHING_EXCAVATION              0
HIGHER_VELOCITY_FLUSHING         0
ANODE_INSTALLED                  0
BREAK_CATEGORIZATION           105
ROADSEGMENTID                    0
STREET                          16
ASSETID                          0
ASSET_SIZE                     121
ASSET_YEAR_INSTALLED           126
ASSET_MATERIAL                 121
ASSET_EXISTS                     0
GLOBALID            

In [28]:
break_data.BREAK_NATURE.value_counts()

UNKNOWN                                      2005
CIRCUMFERENTIAL                               347
CORROSION                                      73
FITTING/JOINT                                  53
LONGITUDINAL                                   29
CIRCUMFERENTIAL AND FITTING/JOINT              24
CORROSION AND CIRCUMFERENTIAL                  15
OTHER                                          14
OTHER: WATER SERVICE                            7
CORROSION AND LONGITUDINAL                      6
CORROSION AND FITTING/JOINT                     5
FITTING/JOINT AND LONGITUDINAL                  2
CORROSION - ROBAR SADDLE CORRODED AT SEAM       1
Name: BREAK_NATURE, dtype: int64

In [29]:
break_data['BREAK_NATURE'] = break_data['BREAK_NATURE'].replace({'OTHER: WATER SERVICE': 'WATER SERVICE'})

In [30]:
break_data.BREAK_NATURE.value_counts()

UNKNOWN                                      2005
CIRCUMFERENTIAL                               347
CORROSION                                      73
FITTING/JOINT                                  53
LONGITUDINAL                                   29
CIRCUMFERENTIAL AND FITTING/JOINT              24
CORROSION AND CIRCUMFERENTIAL                  15
OTHER                                          14
WATER SERVICE                                   7
CORROSION AND LONGITUDINAL                      6
CORROSION AND FITTING/JOINT                     5
FITTING/JOINT AND LONGITUDINAL                  2
CORROSION - ROBAR SADDLE CORRODED AT SEAM       1
Name: BREAK_NATURE, dtype: int64

In [31]:
break_data.BREAK_NATURE.unique()

array(['CORROSION AND FITTING/JOINT', 'UNKNOWN', 'CORROSION',
       'CIRCUMFERENTIAL', 'CORROSION AND CIRCUMFERENTIAL',
       'FITTING/JOINT', 'CIRCUMFERENTIAL AND FITTING/JOINT',
       'LONGITUDINAL', 'WATER SERVICE', 'CORROSION AND LONGITUDINAL',
       'FITTING/JOINT AND LONGITUDINAL', 'OTHER', nan,
       'CORROSION - ROBAR SADDLE CORRODED AT SEAM'], dtype=object)

In [32]:
break_data.BREAK_NATURE.fillna('UNKNOWN', inplace=True)

In [33]:
break_data.BREAK_NATURE.value_counts()

UNKNOWN                                      2094
CIRCUMFERENTIAL                               347
CORROSION                                      73
FITTING/JOINT                                  53
LONGITUDINAL                                   29
CIRCUMFERENTIAL AND FITTING/JOINT              24
CORROSION AND CIRCUMFERENTIAL                  15
OTHER                                          14
WATER SERVICE                                   7
CORROSION AND LONGITUDINAL                      6
CORROSION AND FITTING/JOINT                     5
FITTING/JOINT AND LONGITUDINAL                  2
CORROSION - ROBAR SADDLE CORRODED AT SEAM       1
Name: BREAK_NATURE, dtype: int64

In [34]:
def print_null(df):
    return df.isna().sum()

In [35]:
print_null(break_data)

LONGITUDE                        0
LATITUDE                         0
OBJECTID                         0
WATBREAKINCIDENTID               0
INCIDENT_DATE                    0
BREAK_TYPE                       0
ROAD_CLOSED                      0
SIDEWALK_CLOSED                  0
HOUR_IMPACTED                    0
UNITS_IMPACTED                   0
STATUS                           0
BREAK_NATURE                     0
BREAK_APPARENT_CAUSE             0
POSITIVE_PRESSURE_MAINTANED      0
AIR_GAP_MAINTANED                0
DISINFECTED                      0
MECHANICAL_REMOVAL               0
FLUSHING_EXCAVATION              0
HIGHER_VELOCITY_FLUSHING         0
ANODE_INSTALLED                  0
BREAK_CATEGORIZATION           105
ROADSEGMENTID                    0
STREET                          16
ASSETID                          0
ASSET_SIZE                     121
ASSET_YEAR_INSTALLED           126
ASSET_MATERIAL                 121
ASSET_EXISTS                     0
GLOBALID            

Let's still try and delete some columns that are probably going to be unnecessary in our analysis. Most research papers that I've read so far really only include pipe attributes (length, width, diameter, etc.), last break incident, type of break, number of previous breaks, material of pipe, as well as other data not directly related to the pipe. This includes soil moisture, soil resistivity, soil corrosivity, average temperature, and other attributes like this.

Peaking at the columns again to see which ones we can get rid of...

In [36]:
break_data.columns

Index(['LONGITUDE', 'LATITUDE', 'OBJECTID', 'WATBREAKINCIDENTID',
       'INCIDENT_DATE', 'BREAK_TYPE', 'ROAD_CLOSED', 'SIDEWALK_CLOSED',
       'HOUR_IMPACTED', 'UNITS_IMPACTED', 'STATUS', 'BREAK_NATURE',
       'BREAK_APPARENT_CAUSE', 'POSITIVE_PRESSURE_MAINTANED',
       'AIR_GAP_MAINTANED', 'DISINFECTED', 'MECHANICAL_REMOVAL',
       'FLUSHING_EXCAVATION', 'HIGHER_VELOCITY_FLUSHING', 'ANODE_INSTALLED',
       'BREAK_CATEGORIZATION', 'ROADSEGMENTID', 'STREET', 'ASSETID',
       'ASSET_SIZE', 'ASSET_YEAR_INSTALLED', 'ASSET_MATERIAL', 'ASSET_EXISTS',
       'GLOBALID'],
      dtype='object')

Glancing at these columns again, I'm thinking that 'ROAD_CLOSED', 'SIDEWALK_CLOSED' can be removed. Same with 'DISINFECTED' since I can't really see whether the pipe being disinfected or not contributes to a break, so we can remove that. I don't really want to get rid of any of the ID columns just yet, in case there's some other data that we can cross-reference the IDs of road segments or pipe section IDs or something.

In [37]:
break_data.drop(['ROAD_CLOSED', 'SIDEWALK_CLOSED', 'DISINFECTED'], inplace=True, axis=1)

In [38]:
print_null(break_data)

LONGITUDE                        0
LATITUDE                         0
OBJECTID                         0
WATBREAKINCIDENTID               0
INCIDENT_DATE                    0
BREAK_TYPE                       0
HOUR_IMPACTED                    0
UNITS_IMPACTED                   0
STATUS                           0
BREAK_NATURE                     0
BREAK_APPARENT_CAUSE             0
POSITIVE_PRESSURE_MAINTANED      0
AIR_GAP_MAINTANED                0
MECHANICAL_REMOVAL               0
FLUSHING_EXCAVATION              0
HIGHER_VELOCITY_FLUSHING         0
ANODE_INSTALLED                  0
BREAK_CATEGORIZATION           105
ROADSEGMENTID                    0
STREET                          16
ASSETID                          0
ASSET_SIZE                     121
ASSET_YEAR_INSTALLED           126
ASSET_MATERIAL                 121
ASSET_EXISTS                     0
GLOBALID                         0
dtype: int64

In [39]:
break_data[break_data['BREAK_CATEGORIZATION'].isna()]


LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
2453 -80.460827  43.429301     10561              136193   
2454 -80.505538  43.439726     10562              136194   
2455 -80.474402  43.459717     10563              136195   
2456 -80.468793  43.461157     10564              136196   
2457 -80.486056  43.451335     10565              136197   
...         ...        ...       ...                 ...   
2661 -80.433710  43.434485     46402              144536   
2663 -80.530123  43.424289     46721              144574   
2664 -80.513145  43.425433     46722              144576   
2668 -80.530689  43.441695     48001              144694   
2669 -80.461043  43.439032     48321              144714   

                 INCIDENT_DATE BREAK_TYPE HOUR_IMPACTED UNITS_IMPACTED  \
2453 2019-07-24 00:00:00+00:00       MAIN    8-12 hours              0   
2454 2019-07-26 00:00:00+00:00       MAIN    8-12 hours              0   
2455 2019-07-30 00:00:00+00:00       MAIN    8-12 hours              0   
2456 2019-08-03 00:00:00+00:00       MAIN    8-12 hours              0   
2457 2019-08-14 00:00:00+00:00       MAIN    8-12 hours              0   
...                        ...        ...           ...            ...   
2661 2022-01-05 10:29:04+00:00       MAIN     4-8 hours           0-50   
2663 2022-01-08 09:38:07+00:00       MAIN     4-8 hours           0-50   
2664 2022-01-08 13:31:58+00:00       MAIN     4-8 hours           0-50   
2668 2022-01-18 13:30:27+00:00       MAIN     4-8 hours           0-50   
2669 2022-01-21 06:51:36+00:00       MAIN     4-8 hours           0-50   

                   STATUS  BREAK_NATURE BREAK_APPARENT_CAUSE  \
2453     REPAIR COMPLETED         OTHER                  AGE   
2454     REPAIR COMPLETED         OTHER            CORROSION   
2455     REPAIR COMPLETED         OTHER                  AGE   
2456     REPAIR COMPLETED  LONGITUDINAL                  AGE   
2457     REPAIR COMPLETED         OTHER                  AGE   
...                   ...           ...                  ...   
2661     REPAIR COMPLETED       UNKNOWN              UNKNOWN   
2663     REPAIR COMPLETED       UNKNOWN              UNKNOWN   
2664     REPAIR COMPLETED       UNKNOWN              UNKNOWN   
2668  UNDER INVESTIGATION       UNKNOWN              UNKNOWN   
2669  UNDER INVESTIGATION       UNKNOWN              UNKNOWN   

     POSITIVE_PRESSURE_MAINTANED AIR_GAP_MAINTANED MECHANICAL_REMOVAL  \
2453                           Y                 Y                  Y   
2454                           Y                 Y                  Y   
2455                           Y                 Y                  Y   
2456                           Y                 Y                  Y   
2457                           Y                 Y                  Y   
...                          ...               ...                ...   
2661                           Y                 Y                  Y   
2663                           Y                 Y                  Y   
2664                           Y                 Y                  Y   
2668                           Y                 Y                  Y   
2669                           Y                 Y                  Y   

     FLUSHING_EXCAVATION HIGHER_VELOCITY_FLUSHING ANODE_INSTALLED  \
2453                   Y                        Y               Y   
2454                   Y                        Y               N   
2455                   Y                        Y               Y   
2456                   Y                        Y               Y   
2457                   Y                        Y               Y   
...                  ...                      ...             ...   
2661                   Y                        Y               Y   
2663                   Y                        Y               Y   
2664                   Y                        Y               Y   
2668                   Y                        Y           

In [40]:
break_data.BREAK_CATEGORIZATION.value_counts()

CATEGORY 1    2557
CATEGORY 2       8
Name: BREAK_CATEGORIZATION, dtype: int64

In [41]:
break_data.BREAK_CATEGORIZATION.fillna('UNKNOWN', inplace=True)

In [42]:
break_data.BREAK_CATEGORIZATION.value_counts()

CATEGORY 1    2557
UNKNOWN        105
CATEGORY 2       8
Name: BREAK_CATEGORIZATION, dtype: int64

We looked at 'BREAK_NATURE' earlier in this notebook, but I'm just noticing now that the feature has both 'UNKNOWN' and 'OTHER' as values. We need to convert these to one or the other so everything stays consistent. Let's first see all of the unique values again, and then fix this issue. Since there are a significant amount of the 'UNKNOWN' value we'll most likely convert 'OTHER' to that.

In [43]:
break_data.BREAK_NATURE.unique()

array(['CORROSION AND FITTING/JOINT', 'UNKNOWN', 'CORROSION',
       'CIRCUMFERENTIAL', 'CORROSION AND CIRCUMFERENTIAL',
       'FITTING/JOINT', 'CIRCUMFERENTIAL AND FITTING/JOINT',
       'LONGITUDINAL', 'WATER SERVICE', 'CORROSION AND LONGITUDINAL',
       'FITTING/JOINT AND LONGITUDINAL', 'OTHER',
       'CORROSION - ROBAR SADDLE CORRODED AT SEAM'], dtype=object)

In [44]:
break_data['BREAK_NATURE'] = break_data['BREAK_NATURE'].replace({'OTHER': 'UNKNOWN'})

In [45]:
break_data.BREAK_NATURE.unique()

array(['CORROSION AND FITTING/JOINT', 'UNKNOWN', 'CORROSION',
       'CIRCUMFERENTIAL', 'CORROSION AND CIRCUMFERENTIAL',
       'FITTING/JOINT', 'CIRCUMFERENTIAL AND FITTING/JOINT',
       'LONGITUDINAL', 'WATER SERVICE', 'CORROSION AND LONGITUDINAL',
       'FITTING/JOINT AND LONGITUDINAL',
       'CORROSION - ROBAR SADDLE CORRODED AT SEAM'], dtype=object)

In [46]:
break_data.BREAK_NATURE.value_counts()

UNKNOWN                                      2108
CIRCUMFERENTIAL                               347
CORROSION                                      73
FITTING/JOINT                                  53
LONGITUDINAL                                   29
CIRCUMFERENTIAL AND FITTING/JOINT              24
CORROSION AND CIRCUMFERENTIAL                  15
WATER SERVICE                                   7
CORROSION AND LONGITUDINAL                      6
CORROSION AND FITTING/JOINT                     5
FITTING/JOINT AND LONGITUDINAL                  2
CORROSION - ROBAR SADDLE CORRODED AT SEAM       1
Name: BREAK_NATURE, dtype: int64

In [47]:
print_null(break_data)

LONGITUDE                        0
LATITUDE                         0
OBJECTID                         0
WATBREAKINCIDENTID               0
INCIDENT_DATE                    0
BREAK_TYPE                       0
HOUR_IMPACTED                    0
UNITS_IMPACTED                   0
STATUS                           0
BREAK_NATURE                     0
BREAK_APPARENT_CAUSE             0
POSITIVE_PRESSURE_MAINTANED      0
AIR_GAP_MAINTANED                0
MECHANICAL_REMOVAL               0
FLUSHING_EXCAVATION              0
HIGHER_VELOCITY_FLUSHING         0
ANODE_INSTALLED                  0
BREAK_CATEGORIZATION             0
ROADSEGMENTID                    0
STREET                          16
ASSETID                          0
ASSET_SIZE                     121
ASSET_YEAR_INSTALLED           126
ASSET_MATERIAL                 121
ASSET_EXISTS                     0
GLOBALID                         0
dtype: int64

In [48]:
no_street = break_data.loc[break_data.STREET.isna()]
no_street

LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
155  -80.573004  43.443999      8140                1405   
293  -80.474160  43.471444      8278                 976   
493  -80.427697  43.390080      8478                1978   
693  -80.573004  43.443999      8678                 505   
1443 -80.573004  43.443999      9428                1330   
1607 -80.573004  43.443999      9592                1329   
1616 -80.573004  43.443999      9601                1403   
1758 -80.457321  43.458436      9743                1335   
2039 -80.573004  43.443999     10024                 725   
2052 -80.573004  43.443999     10037                   7   
2086 -80.573004  43.443999     10071                1035   
2102 -80.573004  43.443999     10087                 719   
2152 -80.573004  43.443999     10137                1494   
2595 -80.294712  43.538106     36482              140716   
2596 -80.294712  43.538106     36483              140718   
2597 -80.294712  43.538106     36484              140720   

                 INCIDENT_DATE BREAK_TYPE HOUR_IMPACTED UNITS_IMPACTED  \
155  2009-10-22 00:00:00+00:00       MAIN    8-12 hours              0   
293  2001-03-14 00:00:00+00:00       MAIN    8-12 hours              0   
493  2014-01-10 00:00:00+00:00       MAIN    8-12 hours              0   
693  2011-12-15 15:38:27+00:00       MAIN    8-12 hours              0   
1443 2011-12-15 15:38:28+00:00       MAIN    8-12 hours              0   
1607 2011-12-15 15:38:28+00:00       MAIN    8-12 hours              0   
1616 2009-10-14 00:00:00+00:00       MAIN    8-12 hours              0   
1758 2009-01-10 00:00:00+00:00       MAIN    8-12 hours              0   
2039 2011-12-15 15:38:27+00:00       MAIN    8-12 hours              0   
2052 2011-12-15 15:38:26+00:00       MAIN    8-12 hours              0   
2086 2011-12-15 15:38:27+00:00       MAIN    8-12 hours              0   
2102 2011-12-15 15:38:27+00:00       MAIN    8-12 hours              0   
2152 2010-07-22 00:00:00+00:00       MAIN    8-12 hours              0   
2595 2021-02-17 09:53:29+00:00       MAIN   20-24 hours           0-50   
2596 2021-02-17 10:03:55+00:00       MAIN   20-24 hours           0-50   
2597 2021-02-17 12:05:32+00:00       MAIN   20-24 hours           0-50   

                STATUS     BREAK_NATURE BREAK_APPARENT_CAUSE  \
155   REPAIR COMPLETED          UNKNOWN                OTHER   
293   REPAIR COMPLETED          UNKNOWN                OTHER   
493   REPAIR COMPLETED          UNKNOWN                OTHER   
693   REPAIR COMPLETED          UNKNOWN                OTHER   
1443  REPAIR COMPLETED          UNKNOWN                OTHER   
1607  REPAIR COMPLETED          UNKNOWN                OTHER   
1616  REPAIR COMPLETED          UNKNOWN                OTHER   
1758  REPAIR COMPLETED          UNKNOWN                OTHER   
2039  REPAIR COMPLETED          UNKNOWN                OTHER   
2052  REPAIR COMPLETED          UNKNOWN          COMBINATION   
2086  REPAIR COMPLETED          UNKNOWN                OTHER   
2102  REPAIR COMPLETED          UNKNOWN                OTHER   
2152  REPAIR COMPLETED  CIRCUMFERENTIAL             PRESSURE   
2595         CANCELLED          UNKNOWN              UNKNOWN   
2596         CANCELLED          UNKNOWN              UNKNOWN   
2597         CANCELLED          UNKNOWN              UNKNOWN   

     POSITIVE_PRESSURE_MAINTANED AIR_GAP_MAINTANED MECHANICAL_REMOVAL  \
155                            Y                 Y                  Y   
293                            Y                 Y                  Y   
493                            Y                 Y                  Y   
693                            Y                 Y                  Y   
1443                           Y                 Y                  Y   
1607                           Y                 Y                  Y   
1616                           Y                 Y                  Y   
1758                           Y                 Y   

In [49]:
import plotly
import plotly.express as px

In [50]:
## HIDE TOKEN
from config import token
api_token = token
px.set_mapbox_access_token(token)
fig = px.scatter_mapbox(data_frame=no_street, lat='LATITUDE', lon='LONGITUDE')
fig.update_layout(mapbox_style="carto-positron", mapbox_accesstoken=token)
fig.show();

Let's fill in the `NaN` valued streets with just 'UNKNOWN'

In [51]:
break_data.STREET.fillna('UNKNOWN', inplace=True)

In [52]:
print_null(break_data)

LONGITUDE                        0
LATITUDE                         0
OBJECTID                         0
WATBREAKINCIDENTID               0
INCIDENT_DATE                    0
BREAK_TYPE                       0
HOUR_IMPACTED                    0
UNITS_IMPACTED                   0
STATUS                           0
BREAK_NATURE                     0
BREAK_APPARENT_CAUSE             0
POSITIVE_PRESSURE_MAINTANED      0
AIR_GAP_MAINTANED                0
MECHANICAL_REMOVAL               0
FLUSHING_EXCAVATION              0
HIGHER_VELOCITY_FLUSHING         0
ANODE_INSTALLED                  0
BREAK_CATEGORIZATION             0
ROADSEGMENTID                    0
STREET                           0
ASSETID                          0
ASSET_SIZE                     121
ASSET_YEAR_INSTALLED           126
ASSET_MATERIAL                 121
ASSET_EXISTS                     0
GLOBALID                         0
dtype: int64

In [53]:
break_data.ASSET_SIZE.value_counts()

150.0     1780
300.0      316
200.0      281
100.0       58
450.0       51
600.0       17
250.0       13
25.0        13
50.0         7
1200.0       5
0.0          3
13.0         3
750.0        2
Name: ASSET_SIZE, dtype: int64

So there's 121 missing values for `ASSET_SIZE`. We could simply fill the null values with the mean value, the most frequent value, or just set them to 0. Another way that I feel might be the most appropriate approach is visualize on a map where these asset sizes occur, categorizing by street name, and then seeing if the null values occur on the same streets as some of those already known asset sizes, and then fill in the missing sizes with which neighbours their near.

In [60]:
# pd.set_option("display.max_rows", None)
break_data.loc[break_data.ASSET_SIZE.isna()]

LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
27   -80.435665  43.431450      8012                1113   
43   -80.479748  43.431306      8028                1384   
83   -80.480002  43.431070      8068                 266   
101  -80.486597  43.451541      8086                  35   
110  -80.467894  43.444462      8095                 106   
147  -80.434322  43.432228      8132                1877   
155  -80.573004  43.443999      8140                1405   
184  -80.495594  43.437160      8169                1617   
213  -80.472748  43.478458      8198                1011   
237  -80.497481  43.434806      8222                 528   
244  -80.436342  43.431061      8229                1694   
263  -80.485117  43.450855      8248                1540   
288  -80.435611  43.431482      8273                1798   
420  -80.499501  43.438548      8405                1745   
466  -80.495997  43.436573      8451                1681   
469  -80.475473  43.434025      8454                1196   
472  -80.481642  43.429731      8457                1660   
490  -80.479768  43.431284      8475                1975   
499  -80.486962  43.451686      8484                1985   
512  -80.475181  43.479341      8497                1006   
552  -80.490088  43.445062      8537                1938   
559  -80.433869  43.432608      8544                1945   
565  -80.468841  43.444806      8550                1951   
571  -80.467151  43.444152      8556                1957   
603  -80.477786  43.480303      8588                1008   
660  -80.496431  43.429610      8645                 368   
664  -80.468070  43.444526      8649                2131   
693  -80.573004  43.443999      8678                 505   
719  -80.469802  43.477381      8704                1055   
724  -80.433435  43.432209      8709                 614   
732  -80.436063  43.432909      8717                 514   
738  -80.436450  43.430947      8723                2333   
746  -80.451871  43.426337      8731                2341   
748  -80.479728  43.431328      8733                2152   
753  -80.480505  43.430662      8738                 267   
761  -80.478013  43.480387      8746                1007   
817  -80.479139  43.432708      8802                 128   
839  -80.475498  43.434029      8824                1730   
849  -80.468656  43.444739      8834                 422   
854  -80.474831  43.433927      8839                  60   
857  -80.487137  43.451829      8842                1238   
866  -80.457784  43.441853      8851                1898   
870  -80.485885  43.451202      8855                1903   
920  -80.465449  43.442498      8905                1996   
983  -80.480202  43.448642      8968                1423   
986  -80.436979  43.432027      8971                1174   
1015 -80.458073  43.441873      9000                2036   
1056 -80.479118  43.432772      9041                 127   
1060 -80.479542  43.431585      9045                1383   
1227 -80.479152  43.432663      9212                 126   
1252 -80.435904  43.432763      9237                1533   
1253 -80.527397  43.452220      9238                1318   
1287 -80.483182  43.449985      9272                1275   
1313 -80.436877  43.430472      9298                1428   
1316 -80.497159  43.434917      9301                 529   
1329 -80.475409  43.434016      9314                  59   
1355 -80.479440  43.480775      9340                1180   
1366 -80.496110  43.433060      9351                1661   
1410 -80.434504  43.432122      9395                 516   
1421 -80.475280  43.479377      9406                1010   
1425 -80.496278  43.432295      9410                  68   
1443 -80.573004  43.443999      9428                1330   
1454 -80.435514  43.431538      9439                1188   
1491 -80.479190  43.432526      9476                 129   
1518 -80.434192  43.432306      9503                 517   
1549 -80.495598  43.437140      9534          

Unfortunately 13 out of the previous 16 null values for street names occur in this set. But that's okay for those ones, it just means that we might only end up with 13 null values instead of the current 121. It's better to figure out how to fill in the majority of the values and worry about those few afterwards.

What I will try to do next is look at a sort of breakdown of the asset sizes, listed with their asset ID's and corresponding street names, and see if we can match up any non-null asset sizes from the same street as asset sizes with null values. There's a possibility of imputing the average asset size value from the same street for the null values. For now I am jusy hypothesizing this, I'm not entirely sure if it'll work but why not try right?

In [96]:
asset_size = break_data[['STREET', 'ASSETID', 'ASSET_SIZE']]
# cherry picking null asset sizes to see which streets we coud look at
asset_size.loc[asset_size.ASSET_SIZE.isna()].sample(10)

STREET  ASSETID  ASSET_SIZE
43      OTTAWA ST S    95460         NaN
986        HUBER ST    19990         NaN
1454   FLORENCE AVE    14690         NaN
2086        UNKNOWN        0         NaN
1316     CORAL CRES     8000         NaN
2458     WEBER ST E    90446         NaN
854         MILL ST    26700         NaN
866    FAIRMOUNT RD    83172         NaN
983      WEBER ST E    40880         NaN
2413  SCHWEITZER ST      442         NaN

Next I'll see if there are any matching asset ID's for null and non-null asset sizes. If there are any matches, then it could be safe to say that I could impute the null values with the ID's matching asset size. If this doesn't prove to be true then I'll have to explore more options.

In [86]:
asset_size[asset_size['STREET'] == 'FLORENCE AVE']

STREET  ASSETID  ASSET_SIZE
27    FLORENCE AVE    14690         NaN
147   FLORENCE AVE    14700         NaN
244   FLORENCE AVE    14680         NaN
288   FLORENCE AVE    14690         NaN
738   FLORENCE AVE    14680         NaN
1313  FLORENCE AVE    14680         NaN
1410  FLORENCE AVE    14700         NaN
1454  FLORENCE AVE    14690         NaN
1518  FLORENCE AVE    14700         NaN
1734  FLORENCE AVE    14690         NaN
1803  FLORENCE AVE    14700         NaN
1939  FLORENCE AVE    14690         NaN
2476  FLORENCE AVE    14700         NaN
2486  FLORENCE AVE    14700         NaN

Unfortunately there are no matching ID's for this street. Let's check a few more streets to be sure that this trend might not hold...

In [87]:
asset_size[asset_size['STREET'] == 'OTTAWA ST S']

STREET  ASSETID  ASSET_SIZE
43    OTTAWA ST S    95460         NaN
83    OTTAWA ST S    29660         NaN
395   OTTAWA ST S    64330       450.0
490   OTTAWA ST S    95460         NaN
523   OTTAWA ST S    29630       150.0
596   OTTAWA ST S    29620       150.0
699   OTTAWA ST S    29620       150.0
748   OTTAWA ST S    95460         NaN
753   OTTAWA ST S    29660         NaN
805   OTTAWA ST S    29630       150.0
817   OTTAWA ST S    29680         NaN
926   OTTAWA ST S    29630       150.0
1056  OTTAWA ST S    29680         NaN
1060  OTTAWA ST S    95460         NaN
1227  OTTAWA ST S    29680         NaN
1358  OTTAWA ST S    64330       450.0
1491  OTTAWA ST S    29680         NaN
1572  OTTAWA ST S    29680         NaN
1585  OTTAWA ST S    29630       150.0
1732  OTTAWA ST S    13550       450.0
1851  OTTAWA ST S    29630       150.0
1901  OTTAWA ST S    58280       300.0
2046  OTTAWA ST S    29630       150.0
2050  OTTAWA ST S    29640       150.0
2139  OTTAWA ST S    29630       150.0
2326  OTTAWA ST S    29680         NaN
2345  OTTAWA ST S    29670       150.0
2349  OTTAWA ST S    29680         NaN
2432  OTTAWA ST S    95460         NaN

In [88]:
# display street names with no asset size
streets_with_na = asset_size.STREET[asset_size['ASSET_SIZE'].isna()].unique()

In [89]:
print(streets_with_na)

['FLORENCE AVE' 'OTTAWA ST S' 'WEBER ST E' 'UNKNOWN' 'ST CLAIR AVE'
 'BRIDGE ST E' 'CORAL CRES' 'REX DR' 'NORFOLK CRES' 'MILL ST' 'KEHL ST'
 'HEBEL PL' 'GUERIN AVE' 'OTTAWA ST N' 'STIRLING AVE S' 'WALKER ST'
 'EIGHTH AVE' 'FAIRMOUNT RD' 'HUBER ST' 'GOLFVIEW PL' 'HOFFMAN ST'
 'ANN ST' 'EDWIN ST' 'SCHWEITZER ST']


In [97]:
asset_size[asset_size['STREET'] == 'WEBER ST E']

STREET  ASSETID  ASSET_SIZE
29    WEBER ST E    41030       150.0
101   WEBER ST E    79238         NaN
110   WEBER ST E    40960         NaN
263   WEBER ST E    40820         NaN
310   WEBER ST E    41030       150.0
499   WEBER ST E    40760         NaN
528   WEBER ST E    41030       150.0
565   WEBER ST E    40960         NaN
664   WEBER ST E    40960         NaN
849   WEBER ST E    40960         NaN
857   WEBER ST E    40770         NaN
870   WEBER ST E    40800         NaN
920   WEBER ST E    40980         NaN
983   WEBER ST E    40880         NaN
1180  WEBER ST E    41030       150.0
1221  WEBER ST E    41030       150.0
1287  WEBER ST E    40850         NaN
1428  WEBER ST E    41030       150.0
1552  WEBER ST E    40900         NaN
1666  WEBER ST E    41000       150.0
1707  WEBER ST E    62030       300.0
1784  WEBER ST E    40810         NaN
1903  WEBER ST E    40980         NaN
1961  WEBER ST E    40870         NaN
2183  WEBER ST E    41030       150.0
2259  WEBER ST E    41030       150.0
2335  WEBER ST E    41010       150.0
2367  WEBER ST E    41040       300.0
2430  WEBER ST E    40960         NaN
2457  WEBER ST E    95308         NaN
2458  WEBER ST E    90446         NaN
2459  WEBER ST E    95014         NaN
2460  WEBER ST E    40790         NaN
2491  WEBER ST E    41000       150.0
2496  WEBER ST E    41030       150.0
2592  WEBER ST E    41030       150.0
2669  WEBER ST E    41030       150.0

Well there we have it, each ID is unique to the asset size whether it's missing or not.

I will fill in the `NaN` values with the most frequent number (mode) for now.

In [101]:
asset_size.ASSET_SIZE.value_counts()

150.0     1780
300.0      316
200.0      281
100.0       58
450.0       51
600.0       17
250.0       13
25.0        13
50.0         7
1200.0       5
0.0          3
13.0         3
750.0        2
Name: ASSET_SIZE, dtype: int64

In [102]:
# fill asset size with the mode of the column
break_data['ASSET_SIZE'].fillna(break_data['ASSET_SIZE'].mode()[0], inplace=True)

In [103]:
break_data.ASSET_SIZE.value_counts()

150.0     1901
300.0      316
200.0      281
100.0       58
450.0       51
600.0       17
250.0       13
25.0        13
50.0         7
1200.0       5
0.0          3
13.0         3
750.0        2
Name: ASSET_SIZE, dtype: int64

In [104]:
print_null(break_data)

LONGITUDE                        0
LATITUDE                         0
OBJECTID                         0
WATBREAKINCIDENTID               0
INCIDENT_DATE                    0
BREAK_TYPE                       0
HOUR_IMPACTED                    0
UNITS_IMPACTED                   0
STATUS                           0
BREAK_NATURE                     0
BREAK_APPARENT_CAUSE             0
POSITIVE_PRESSURE_MAINTANED      0
AIR_GAP_MAINTANED                0
MECHANICAL_REMOVAL               0
FLUSHING_EXCAVATION              0
HIGHER_VELOCITY_FLUSHING         0
ANODE_INSTALLED                  0
BREAK_CATEGORIZATION             0
ROADSEGMENTID                    0
STREET                           0
ASSETID                          0
ASSET_SIZE                       0
ASSET_YEAR_INSTALLED           126
ASSET_MATERIAL                 121
ASSET_EXISTS                     0
GLOBALID                         0
dtype: int64

In [107]:
break_data.ASSET_YEAR_INSTALLED.value_counts()

1958.0    158
1966.0    155
1967.0    153
1962.0    105
1953.0    105
1960.0     94
1959.0     86
1964.0     86
1961.0     84
1954.0     79
1965.0     78
1952.0     74
1963.0     69
1957.0     66
1969.0     58
1968.0     54
1974.0     53
2017.0     51
1955.0     51
2016.0     51
1956.0     47
2008.0     46
2010.0     45
1971.0     33
2015.0     30
1970.0     28
1976.0     26
2018.0     25
1949.0     24
1950.0     23
1948.0     21
1981.0     20
2014.0     20
1972.0     20
1951.0     18
1978.0     17
1973.0     17
1975.0     17
2012.0     16
1979.0     15
1937.0     14
1947.0     13
2013.0     13
2011.0     13
1922.0     13
1940.0     12
1995.0     12
1988.0     12
1925.0     11
2000.0     11
1977.0     11
2009.0     11
1946.0     11
1987.0     10
2001.0     10
1923.0     10
1980.0      9
1889.0      9
1989.0      7
1999.0      7
1924.0      7
2006.0      7
2005.0      6
1994.0      6
1933.0      6
1992.0      6
1935.0      5
2004.0      5
1991.0      5
1930.0      5
2002.0      4
1983.0

In [108]:
break_data[break_data['ASSET_YEAR_INSTALLED'].isna()].sample(10)

LONGITUDE   LATITUDE  OBJECTID  WATBREAKINCIDENTID  \
2349 -80.478877  43.433524     10334                2355   
1803 -80.434770  43.431969      9788                 518   
27   -80.435665  43.431450      8012                1113   
1491 -80.479190  43.432526      9476                 129   
1410 -80.434504  43.432122      9395                 516   
552  -80.490088  43.445062      8537                1938   
237  -80.497481  43.434806      8222                 528   
472  -80.481642  43.429731      8457                1660   
2458 -80.486062  43.451360     10566              136198   
1789 -80.435535  43.432423      9774                 513   

                 INCIDENT_DATE BREAK_TYPE HOUR_IMPACTED UNITS_IMPACTED  \
2349 2018-01-07 00:00:00+00:00       MAIN    8-12 hours              0   
1803 2001-02-26 00:00:00+00:00       MAIN    8-12 hours              0   
27   2007-03-07 00:00:00+00:00       MAIN    8-12 hours              0   
1491 2000-01-12 00:00:00+00:00       MAIN    8-12 hours              0   
1410 1997-01-25 00:00:00+00:00       MAIN    8-12 hours              0   
552  2014-02-07 00:00:00+00:00       MAIN    8-12 hours              0   
237  1999-03-08 00:00:00+00:00       MAIN    8-12 hours              0   
472  2011-12-15 15:38:26+00:00       MAIN    8-12 hours              0   
2458 2019-08-14 00:00:00+00:00       MAIN    8-12 hours              0   
1789 2003-02-19 00:00:00+00:00       MAIN    8-12 hours              0   

                STATUS BREAK_NATURE BREAK_APPARENT_CAUSE  \
2349  REPAIR COMPLETED      UNKNOWN                OTHER   
1803  REPAIR COMPLETED      UNKNOWN                OTHER   
27    REPAIR COMPLETED      UNKNOWN                OTHER   
1491  REPAIR COMPLETED      UNKNOWN                OTHER   
1410  REPAIR COMPLETED      UNKNOWN                OTHER   
552   REPAIR COMPLETED      UNKNOWN                OTHER   
237   REPAIR COMPLETED      UNKNOWN                OTHER   
472   REPAIR COMPLETED      UNKNOWN                OTHER   
2458  REPAIR COMPLETED      UNKNOWN                  AGE   
1789  REPAIR COMPLETED      UNKNOWN                OTHER   

     POSITIVE_PRESSURE_MAINTANED AIR_GAP_MAINTANED MECHANICAL_REMOVAL  \
2349                           Y                 Y                  Y   
1803                           Y                 Y                  Y   
27                             Y                 Y                  Y   
1491                           Y                 Y                  Y   
1410                           Y                 Y                  Y   
552                            Y                 Y                  Y   
237                            Y                 Y                  Y   
472                            Y                 Y                  Y   
2458                           Y                 Y                  Y   
1789                           Y                 Y                  Y   

     FLUSHING_EXCAVATION HIGHER_VELOCITY_FLUSHING ANODE_INSTALLED  \
2349                   Y                        Y               Y   
1803                   Y                        Y               Y   
27                     Y                        Y               Y   
1491                   Y                        Y               Y   
1410                   Y                        Y               Y   
552                    Y                        Y               Y   
237                    Y                        Y               Y   
472                    Y                        Y               Y   
2458                   Y                        Y               Y   
1789                   Y                        Y               Y   

     BREAK_CATEGORIZATION  ROADSEGMENTID        STREET  ASSETID  ASSET_SIZE  \
2349           CATEGORY 1          10877   OTTAWA ST S    29680       150.0   
1803           CATEGORY 1          12094  FLORENCE AVE    14700       150.0   
27             CATEGORY 1          12095  FLORENCE AVE    14690     

From reading different studies, it seems as though the year the pipe was installed or rather the age of the pipe is a critical factor in predicting breaks/time of failure. In this case, I believe I may have to drop the observations where no year is indicated. 

In [109]:
break_data.dropna(axis=0, subset=['ASSET_YEAR_INSTALLED'], inplace=True)

In [111]:
print(break_data.shape)
print_null(break_data)

(2544, 26)


LONGITUDE                      0
LATITUDE                       0
OBJECTID                       0
WATBREAKINCIDENTID             0
INCIDENT_DATE                  0
BREAK_TYPE                     0
HOUR_IMPACTED                  0
UNITS_IMPACTED                 0
STATUS                         0
BREAK_NATURE                   0
BREAK_APPARENT_CAUSE           0
POSITIVE_PRESSURE_MAINTANED    0
AIR_GAP_MAINTANED              0
MECHANICAL_REMOVAL             0
FLUSHING_EXCAVATION            0
HIGHER_VELOCITY_FLUSHING       0
ANODE_INSTALLED                0
BREAK_CATEGORIZATION           0
ROADSEGMENTID                  0
STREET                         0
ASSETID                        0
ASSET_SIZE                     0
ASSET_YEAR_INSTALLED           0
ASSET_MATERIAL                 0
ASSET_EXISTS                   0
GLOBALID                       0
dtype: int64

Lucky for us that took care of our missing values in the `ASSET_MATERIAL` column. Now we have a clean dataset finally to do some EDA and feature engineering.

In [112]:
break_data.to_csv('./data/processed/cleaned_break_data.csv', index=False)